File: src/main/java/dfs/TopkInBinaryTree.java
package dfs;

import structure.BinaryTree;
import structure.BinaryTree.TreeNode;

public class TopkInBinaryTree {
    public static void main(String args[]){
        Integer[] arrs = {8, 5, 12, 3, 10, 11, 15};
        BinaryTree binaryTree = new BinaryTree(arrs);

        System.out.println(kthLargest(binaryTree.root, 1));
    }

    static int count;
    static int res;

    /**
     * 右中左递归查找第k大元素值（从大到小遍历）
     * 时间复杂度：O(N)，会遍历完树中每一个元素，与k无关
     */
    public static int kthLargest(TreeNode root, int k) {
        count = k;
        kthLargest(root);
        return res;
    }

    public static void kthLargest(TreeNode root){
        if (root != null){ // 非初始状态，开始往子节点遍历
            System.out.println(root.val+" 入栈");
            kthLargest(root.right); // 先往右节点走
            if (count == 1){
                res = root.val;
                System.out.println("res = "+res);
                System.out.println(root.val+" 是第 "+count+" 大元素");
                count --;
                System.out.println(root.val+" 出栈");
                return;
            }
            System.out.println(root.val+" 是第 "+count+" 大元素");
            count --;
            kthLargest(root.left); // 再往左边走
            System.out.println(root.val+" 出栈");
        }
        // root == null 时为初始状态，直接return
    }
}

File: CppAlgorithms/dfs/TopkInBinaryTree.cpp
#include <stddef.h>
#include <vector>
#include <iostream>
#include "../structure/TreeNode.h"
using namespace std;
#define INF 0xfffffff

/**
 * https://leetcode-cn.com/problems/er-cha-sou-suo-shu-de-di-kda-jie-dian-lcof/
 * 给定一棵二叉搜索树，请找出其中第k大的节点。
 * 输入: root = [3,1,4,null,2], k = 1
 *    3
 *   / \
 *  1   4
 *   \
 *    2
 * 输出: 4
 */
class TopkInBinaryTree {
private:
    void dfs(TreeNode* node, vector<int> &v){
        if (node == NULL)
            return;
        dfs(node->left, v);
        v.push_back(node->val);
        dfs(node->right, v);
    }
public:
    int kthLargest(TreeNode* root, int k) {
        vector<int> v;
        dfs(root, v);
        return v[v.size() - k];
    }
};

int main() {
    vector<int> v{3,1,4,INF,2};
    BinaryTree tree(v);
    TopkInBinaryTree s;
    int res = s.kthLargest(tree.root, 2);
    cout << res << endl;
    return 0;
}